### ISSUE TITLE
win_nssm: app_parameters - Unrecognized escape sequence

### ISSUE THREAD

**Issue by dreamcat4** [ansible/ansible-modules-extras#1195]
Hi,
the following task:

``` yaml
- win_nssm:
    app_parameters:
      "-document_root": "C:\\ProgramData\\wsusoffline\\iso"
```

Results in this error message:

``` sh
failed: [ga-z170x-g3] => {"changed": true, "failed": true}
msg: parsing "C:\ProgramData\wsusoffline\iso" - Unrecognized escape sequence \P.

FATAL: all hosts have already failed -- aborting
```

It seems the `\\` escaping is wrong in your module. Because when I put in an extra 2 `\\` it then works:

``` sh
- win_nssm:
    app_parameters:
      "-document_root": "C:\\\\ProgramData\\\\wsusoffline\\\\iso"
```

The problem is that we can't do that (make 4 backslashes) for templated parameters which may also be used elsewhere in ansible. The level of escaping must be consistent throughout ansible. So therefore I beleive there is a bug somewhere in the arguments parsing of `app_parameters` hash of this module `win_nssm`.

**Comment by dreamcat4** [comment]
@gfrank If you could please take a look at this issue when you have the time.

**Comment by georgefrank** [comment]
sure thing!

**Comment by georgefrank** [comment]
PR made https://github.com/ansible/ansible-modules-extras/pull/1198

**Comment by dreamcat4** [comment]
Thanks man! I'll try it out.

**Comment by dreamcat4** [comment]
BTW (not very important / doesn't matter) I just also noticed in the other file `win_nssm.py` you had copy-pasted `password:` declaration twice by accident. (from your 13 days ago documentation change).

Back to testing the escaping.

**Comment by georgefrank** [comment]
Oh noes! I can tackle that now too.

**Comment by georgefrank** [comment]
Since you're using the win_nssm module - what do you think of the whole app_parameters implementation? Specifically that it uses an underscore to denote a non-key-value parameter (`"_":"foo"`). It seems clunky to me, but the one thing it has going for it is that the order of all parameters is maintained when the underlying nssm command is created.

I was thinking of breaking key-value params and single-argument params into two separate variables, but then single-argument params would always be appended to the underlying nssm command before key-value params (with each list maintaining order). I'm not sure that this would matter much in practice with nssm since the functionality that's exposed through ansible is limited to service create/start/stop/delete, and it could make for a more intuitive ansible task.
Any comments would be appreciated!

**Comment by dreamcat4** [comment]
OK the escaping seems not to give any error for me now. With the escaping either way before or after. So it seems good to me.

In regards to `app_parameters:` I think it's fine / very good because you clearly documented it, including how to put un-named `args` using `_:`. So I don't mind myself.

However if you want to change that behaviour I would recommend a single array of quoted strings e.g. `[ "arg1", "-c", "switch-value", "arg2","--switch2=foo","etc..."]`. Which would be 1 variable for the argv[] verbatim (no splitting into 2 variables). And the arguments order is still maintained that way...

However array method has it's own drawback. The advantage of using hash (as you currently do) is a simple merge to override named parameters (with the same key name). Which is also of some value in SCCM environments.

If anything I wouldn't go out of your way too much as your current solution seems to works very well IMHO. But if you could test if the value of `app_parameters:` value is either a list or a dict then users could be free to choose either way depending if the bulk of their arguments are primarily named `-switches` or un-named `args`.

**Comment by georgefrank** [comment]
Thanks! Since it works now and is documented well enough, I'll just let it be.

**Comment by dreamcat4** [comment]
@georgefrank Hi again!

You fixed the escaping but there seem some other bug now (in latest version) with the `app_parameters`. I didn't have your very latest version before, but it had been working.

I'll explain the issue, and how to reproduce it:

``` yaml
- name: Windows Service - NSSM Create and Start - Mongoose HTTP Server
  win_nssm:
    name: mongoose_wsusoffline_iso
    application: "{{ mongoose_bin_win }}"
    app_parameters:
      "-document_root": "{{ wsusoffline_win }}\\iso"
      "-listening_port": "35350"
    state: restarted
```

That's my task ^^ in ansible. Run that, then go to windows machine, and type `nssm edit mongoose_wsusoffline_iso` to bring up the NSSM Service editing window.

In that windows, look in `Arguments` text input field... It will show as if the arguments are now being concactenated togther without any spaces. But actually if you copy-paste that string to a text editor then there seems to be a `\n` newline character still in there:

``` sh
-document_root C:\ProgramData\wsusoffline\iso
-listening_port=35350
```

Perhaps it should be a space ` ` instead? This is causing my webserver not to start up correctly.

Also for some reason the 2nd parameter mysteriously has an equals `=` sign between it... But we didn't specify that in our above ansible task yaml declaration. Again, a space ` `would be nice. As we can always do an equals sign with`_: arg=value` instead / if necessary.

**Comment by georgefrank** [comment]
Shame on me! On it now.

**Comment by robynbergeron** [comment]
Hi @dreamcat4 -- thanks for filing this bug report, and @georgefrank -- thanks for your quick response. Looks like you guys have things underway here, just adding labels :) Carry on! <3 :)

**Comment by robynbergeron** [comment]
[module: windows/win_nssm.py]

**Comment by georgefrank** [comment]
Heyo! I tried to get ahead of myself - using the [Regex]::Escape(), when really I only needed to escape the slashes. https://github.com/ansible/ansible-modules-extras/pull/1198 has the updated code now.

**Comment by dreamcat4** [comment]
OK I test again and report back.

**Comment by dreamcat4** [comment]
@georgefrank Thanks it worked fine the app params issue fixed now and is all OK as far as I can see from here. Thanks again for this module. It's very good one.

Please merge code when you can.

**Comment by robynbergeron** [comment]
@dreamcat4 & @georgefrank -- I just marked #1198 for inclusion, as FYI. Will close this out after it gets merged. 

Thanks, everyone!

**Comment by nitzmahone** [comment]
fix merged from #1198
